### ISSUE TITLE
Events in libui-ng?

### ISSUE THREAD

**Issue by rubyFeedback** [libui-ng/libui-ng#131]
Hey folks,

There may not be a possibility for this suggestion but I wanted to suggest it nonetheless.

Yesterday I wrote a quick and dirty ruby-gtk3 shell interface to my commandline shell
written in ruby. This is super-hackish, written by a hobbyist with only average skills, 
and has tons of bugs - but, it also works for some use cases.

The ruby-gtk3 interface currently looks like this:

https://i.imgur.com/Vb0pySJ.png

This is not super-pretty but the functionality works. I input commands in the first
entry, and the text-view buffer below tells me what happens. At a later point
I may improve this but for now I wanted to have a shell and a terminal in one,
and that works on ruby-gtk3 on linux.

On windows this is different; getting gtk3 to work there is annoying. I hate that.

But libui works as-is thanks to kojix2 writing the bindings.

I had to cut down tons of features though. One thing that annoys me is that I
have to use an entry where I type something but then I have to click on a separate
button. On the ruby-gtk3 entry I hit the "enter" key as-is, while I am inside of
the entry, and then it is evaluated.

Now: I have a work around. I can periodically scan the text of the entry, and
if it changed, I can decide whether I "fire" the event or not. So I can check every
0.3 seconds or so and this is very similar to hitting enter key. But not very
elegant.

So my question is: would it be possible to support SOME events for libui?

I am not asking for all events, but a simple one would be for these kind
of checks:

"if button press event is ENTER key,
    respond to this"

I don't know how difficult this is on different platforms. Perhaps it could
be implemented for only one platform. (Ok this may not work, but actually
I use libui primarily for windows only. On linux I can use ruby-gtk or just
the commandline, but on windows I like libui since it cuts down development
time. I can use java + swing or JavaFX and that works too but java is
so verbose and takes significantly longer to write the code in than ruby).

Anyway. I just wanted to mention this here. Perhaps others may also want
events and someone understands C/C++ sufficiently well to add a first
prototype or something. Then we could perhaps extend from this point
eventually until more events would be available.

**Comment by AndyObtiva** [comment]
This already has a ticket over here:
https://github.com/andlabs/libui/issues/325

SWT supports that already. You have common keyboard events in all widgets. 

See all the addXyzListener methods here:

https://help.eclipse.org/latest/index.jsp?topic=/org.eclipse.platform.doc.isv/reference/api/org/eclipse/swt/widgets/package-summary.html

**Comment by szanni** [comment]
These are a lot of different ideas/requests mangled together.

Keyboard events are somewhat present in uiArea, but nothing generic, applicable to all controls as far as I can tell. I would say, this is definitely a desired feature. We just need someone to implement this in a more generic way.

As for your particular instance, why not use the `uiEntryOnChanged()` callback instead of having the user press enter?

Regarding fonts and colors: sure it would be great to have a simpler API, some wrapper for uiArea, if you are only interested in displaying information.
But something akin to `widget.set_font` looks technically challenging and IMO is contrary to what I, as the application user, actually want. I want my button and all generic widgets to use the system font in the size the system specifies. This is a usability issue for me. A text editor or type setting program is different and uiAttributedString looks like a reasonable API to me for an extremely complex problem.

**Comment by AndyObtiva** [comment]
I just added this statement to my last comment:

“ in any case, I agree that styling all widgets via font, foreground colour, background colour, image background, and vector graphics would be great. SWT also supports that already (though only certain widgets like label can have an image background, but all can have arbitrarily painted vector graphics on top of the native widget rendering)”

I think it’s important to look into competing solutions like SWT, and if they’ve already solved a problem, like allowing customization of fonts on all widgets, libui should not waste time arguing details, yet instead unashamedly copy the solution.

**Comment by szanni** [comment]
> I think it’s important to look into competing solutions like SWT

Definitely, and this is being done on a constant basis.

> like allowing customization of fonts on all controls (useful for elderly accessability requirements)

I would argue that each platform handles such matters a _lot better_ and _more consistent_ than any application developer is capable of. Not to say that such an API is undesirable in general, but I would not tout it as an accessibility feature. Quite the opposite actually.

> libui should not waste time arguing details, yet instead unashamedly copy the solution.

Pull requests are welcome.

Apart from that I would not say that discussing features is a waste of time. There are many things to consider, like: How complex is the solution? Who is going to maintain it? Does it align with the goals of the library: `Simple and portable`? etc.

**Comment by cody271** [comment]
> 
> > like allowing customization of fonts on all controls (useful for elderly accessability requirements)
> 
> I would argue that each platform handles such matters a _lot better_ and _more consistent_ than any application developer is capable of. Not to say that such an API is undesirable in general, but I would not tout it as an accessibility feature. Quite the opposite actually.
>

Yes the answer is to use your platform's native accessibility stack. Setting custom fonts on controls for text magnification and custom colors for high contrast are definitely important accessibility features. These are both builtin accessibility features to every major platform.

Reimplementing them inside your app would make the situation worse and not better though, as the only reason to do that usually is in self drawn toolkits where you don't have a choice.. this is the benefit of being totally native: we get these sorts of features for free! Unless I'm missing something here?

**Comment by cody271** [comment]
> in any case, I agree that styling all controls via font, foreground colour, background colour, image background, and vector graphics would be great. SWT also [supports that already](https://help.eclipse.org/latest/topic/org.eclipse.platform.doc.isv/reference/api/org/eclipse/swt/widgets/Control.html) (though only certain controls like label can have an image background, but all can have arbitrarily painted vector graphics on top of the native control rendering)

Definitely supportive of adding a custom theming API, for the edge cases where this is required.
Pull requests are welcome! See original discussion: [here](https://github.com/andlabs/libui/issues/313).

**Comment by AndyObtiva** [comment]
As a very experienced user of SWT, I’d say having customizations to font, foreground, background, etc… is indispensable. In software teams I’ve been, we’ve used them in teams at businesses to white label (brand) apps (even while using native controls). And, I’ve personally used for personal and open source app projects too. 

As such, I had the best of both worlds. I could be fully native in certain cases and I can customize when needed. 

I’m looking forward to LibUI supporting this flexibility eventually so that I can use to build apps that start faster than JVM apps.
